src/orion/core/worker/warm_start/multi_task_wrapper.py

```diff
@@ -0,0 +1,245 @@
+"""Wrapper that makes an algorithm warm-starteable by observing trials from different tasks.
+
+The wrapper adds a `task_id` dimension to the search space, and uses the value of `task_id` to
+differentiate between the current "target" experiment (`task_id==0`) and the other experiments
+from the knowledge base (`task_id>0`).
+
+The trials from the knowledge base that can fit within the same space are reused and given a
+different value for a new "task_id" dimension.
+"""
+from __future__ import annotations
+
+import copy
+from collections import defaultdict
+from contextlib import contextmanager
+from logging import getLogger
+from typing import Iterable, TypeVar
+
+from orion.algo.base import BaseAlgorithm
+from orion.algo.space import Categorical, Space
+from orion.core.utils.format_trials import dict_to_trial
+from orion.core.worker.algo_wrappers.algo_wrapper import (
+    AlgoWrapper,
+    _copy_status_and_results,
+)
+from orion.core.worker.warm_start.knowledge_base import ExperimentInfo, KnowledgeBase
+from orion.core.worker.warm_start.warm_starteable import (
+    WarmStarteable,
+)
+from orion.core.worker.trial import Trial
+
+AlgoType = TypeVar("AlgoType", bound=BaseAlgorithm)
+log = getLogger(__file__)
+
+
+class MultiTaskWrapper(AlgoWrapper[AlgoType], WarmStarteable):
+    """Wrapper that makes the algo "multi-task" by adding a task id to the inputs."""
+
+    def __init__(
+        self,
+        space: Space,
+        algorithm: AlgoType,
+    ):
+        log.debug(
+            f"Creating a MultiTaskAlgo wrapper: space {space}, algo config: {algorithm.configuration}"
+        )
+        self.current_task_id: int = 0
+
+        # TODO: IDEA: Use `self.registry` only for the current (target) task, and let `self.algorithm.registry`
+        # contain trials from all tasks, with task IDs added, including the target task?
+
+    def transform(self, trial: Trial) -> Trial:
+        return self._add_task_id(trial, self.current_task_id)
+
+    def reverse_transform(self, trial: Trial) -> Trial:
+        return self._remove_task_id(trial)
+
+    @classmethod
+    def transform_space(cls, space: Space, knowledge_base: KnowledgeBase) -> Space:
+        """Transform the space, so that the algorithm that is passed to the constructor is already
+        transformed.
+        """
+        # TODO: Should we have this dimension be larger than the current number of
```

Comment:
I agree that everything should be 100% reproducible. Is this in response to the TODO here though? If so, I don't understand the link.

Reply:
This would make the dimension size (number of categories) fix.

Reply:
> Perhaps we should make the KB specification so that running the experiment twice would query exactly the same experiments.

I still don't quite understand what you mean here.
The number of categories would be fixed, if and only if the user creates the same knowledge base, with the same experiments, correct? Sure, the ordering will be deterministic (e.g. sorted by name or similarity or something).

Also, about the idea of limiting the KB specification, do you mean that, after creating a KB with a storage object A, if new experiments are added to A, and then a new KB is created with Storage A, then it should have the same "view" of experiments/trials as the first KB? If so, I disagree.

Reply:
If a new experiment is added, then the number of tasks will increase, thus there will be one additional category for the task_id dimension.

With respect to freezing the KB, from what I remember we agreed offline that this should not be implemented and users should rather rely on the upcoming feature to checkpoint a storage which they could assume would not change if they don't make any modifications to it. Maybe I misunderstood and we had something different in mind?

Reply:
No that's all good.